Test for posix lib + newlib (also start to fix issues)

galak opened the pull request.

galak committed:
lib: posix: Fix compile error with mqueue.h

With newer newlib we get a build error with mqueue.h realted to mode_t.
Let's just let newlib define mode_t and have minimal libc also define
it in sys/types.h.  So we remove the duplicated definition in
posix/unistd.h.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
tests: posix: add newlib cfg option

Add a build variant of the posix tests with newlib enabled.  Since the
interation of posix is greatly related to the libc we should have these
tests build with and without newlib support.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak commented:
@vanti I think this might be a cleaner test addition for newlib and posix+lib to get us more coverage.

pfalcon commented on include/posix/mqueue.h:
```diff
@@ -17,7 +17,10 @@ extern "C" {
 
 typedef void *mqd_t;
+#ifndef _MODE_T_DECLARED
 typedef unsigned int mode_t;
```
Well, it's implausible situation that mode_t is defined by mqueue.h. Per POSIX, it's defined by sys/types.h: http://pubs.opengroup.org/onlinepubs/9699919799/basedefs/sys_types.h.html . So, let's start to actually fix these, not work around. I.e., should be added to lib/libc/minimal/include/sys/types.h for minimal libc case, it's of course already there for newlib, and should be removed from include/posix/*.

galak replied:
What's interesting is we also define `mode_t` in `include/posix/unistd.h`

pfalcon replied:
All POSIX headers is complete mess. And we either clean them up and rely on newlib headers, or choose the hard way of cleaning them up **and** recreating newlib headers for minimal libc ;-).

galak replied:
We'll I cleaned it up so mode_t comes from newlib or minlibc types.h

vanti commented:
> @vanti I think this might be a cleaner test addition for newlib and posix+lib to get us more coverage. Take a look and see if you agree and we can close out your PR.

@galak I am not too familiar with how the testcase.yaml files work, but when I copy your changes into my local repo and rebuild the posix/fs and posix/common tests I do not see the build error from sdk-ng#45. Adding `CONFIG_NEWLIB_LIBC=y` to posix/fs/prj.conf does the trick though.

pfalcon reviewed (approved):
Thanks. I won't be surprised if the build now will fail because &lt;sys/types.h> is not included somewhere, but that's the way to resolve such issues.

pfalcon commented:
Ok, so the typical situation, you make one change, there's a cascade of fallouts: https://app.shippable.com/github/zephyrproject-rtos/zephyr/runs/35304/2/tests .

In https://github.com/zephyrproject-rtos/zephyr/pull/12984 , I ended up banning native_posix, riscv32, xtensa, qemu_x86_64 (i.e. effectively everything but x86 and arm I guess ;-) ), and it still has a hundred of failures on last check (but it has massive changes of course).

galak committed:
lib: posix: Fix compile error with time.h and older newlib

Some older variants of newlib (like what we have on xtensa or riscv in
SDK 0.9.5) define timespec and itimerspec in sys/types.h.  The timespec
can be handled by a check of __timespec_defined.  However itimerspec
doesn't have anything similar so we have to do it by newlib version
info.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

pfalcon commented on include/posix/time.h:
```diff
@@ -10,15 +10,29 @@
 };
+#endif
 
+/* Older newlib's like 2.{0-2}.0 don't define any newlib version defines, only
```
@galak, The proper fix for this would be the same as for mode_t - figure out the proper header belongs to, rely on newlib's existing definition there, and move def from posix/* to minimal/*. (Again, when we repeat that process a hundred of times, we'll find that we just recreated newlib headers in minimal. So, it's still not to late to say "POSIX and minimal don't plat well together, a [decent] libc is part of POSIX", and concentrate only on POSIX+newlib).

nashif commented:
@galak Why was this merged with CI failing? Now everything is failing in CI because of this.
